core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java

```diff
@@ -1,20 +1,59 @@
 package io.quarkus.runtime;
 
-import java.lang.reflect.Method;
+import static io.quarkus.runtime.serial.Deserializer.deserializeResource;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.UndeclaredThrowableException;
+
+import io.quarkus.runtime.execution.Execution;
+import io.quarkus.runtime.execution.ExecutionChain;
+import io.quarkus.runtime.execution.ExecutionContext;
+import io.quarkus.runtime.execution.RestartingExecutionHandler;
+import io.quarkus.runtime.execution.ShutdownHookExecutionHandler;
+import io.quarkus.runtime.execution.SignalHandlerExecutionHandler;
 
 /**
- * The main entry point class, calling main allows you to bootstrap quarkus
- * <p>
- * Note that at native image generation time this is replaced by {@link io.quarkus.runtime.graal.QuarkusReplacement}
- * which will avoid the need for reflection.
- * <p>
- * TODO: how do we deal with static init
+ * The main entry point class of Quarkus.
  */
 public class Quarkus {
 
-    public static void main(String... args) throws Exception {
-        Class main = Class.forName("io.quarkus.runner.GeneratedMain");
-        Method mainMethod = main.getDeclaredMethod("main", String[].class);
-        mainMethod.invoke(null, (Object) args);
+    static {
+        Timing.staticInitStarted();
+        try {
+            final Class<?> init = Class.forName("io.quarkus.runtime.generated.Init", true, Quarkus.class.getClassLoader());
+            INITIAL_CONTEXT = (ExecutionContext) init.getDeclaredMethod("getInitialContext").invoke(null);
+        } catch (ClassNotFoundException e) {
+            throw new NoClassDefFoundError(e.getMessage());
+        } catch (IllegalAccessException e) {
+            throw new IllegalAccessError(e.getMessage());
+        } catch (InvocationTargetException e) {
+            try {
+                throw e.getCause();
+            } catch (Error | RuntimeException e2) {
+                throw e2;
+            } catch (Throwable throwable) {
+                throw new UndeclaredThrowableException(throwable);
+            }
+        } catch (NoSuchMethodException e) {
+            throw new NoSuchMethodError(e.getMessage());
+        }
+        STARTUP_CHAIN = deserializeResource(ExecutionChain.class, "META-INF/quarkus.init");
```

Comment:
> Even though this will not increase boot time on Substrate the JVM is still very much a first class citizen. JVM bootstrap is just as important as substrate bootstrap.

This approach has one key advantage: it is expedient to implement.  Of course other materialization approaches are possible as well; we can probably follow this up with an enhancement to bytecode-generate the chain materialization.

Reply:
Just because it is 'expedient to implement' I still don't understand how it is an improvement on what we currently have or why any of this is necessary for command mode. Basically these changes do lots of things in a different way, but different is not the same as better.

It looks like most of the changes come from the idea that arbitrary stages of the application might need to be restarted, but I don't really think this is relevant to command mode, and I don't think that it is a good idea in general.

Reply:
> Just because it is 'expedient to implement' I still don't understand how it is an improvement on what we currently have or why any of this is necessary for command mode. Basically these changes do lots of things in a different way, but different is not the same as better.

Different is not the same as worse either.  It cleanly solves problems that the original approach cannot provide clean solutions for, like having a single exit point and correctly handling `System.exit()`. 

> It looks like most of the changes come from the idea that arbitrary stages of the application might need to be restarted, but I don't really think this is relevant to command mode, and I don't think that it is a good idea in general.

Command mode and other things need to intercept startup at various points.  They may need to exit early.  Being able to restart is a natural consequence of this.  If the program does not abide by the contracts set out in the API then it's faulty, and it's not only restart that will likely show this to be the case; if it does, then it's restartable and there's nothing to worry about.  I'm sorry you don't think it's a good idea, but frankly I'm not responsible for your feelings.  I'm happy to answer your questions about the code though.